sklearn/compose/_column_transformer.py

```diff
@@ -903,6 +904,11 @@ def _check_X(X):
     """Use check_array only on lists and other non-array-likes / sparse"""
     if hasattr(X, "__array__") or sparse.issparse(X):
         return X
+    elif hasattr(X, "__dataframe__"):
+        with suppress(ImportError):
+            from pandas.api.interchange import from_dataframe
+
+            return from_dataframe(X)
```

Comment:
this looks kinda dirty to me. Shouldn't we instead have something in `_validate_data`  or `check_X_y, ...` to convert to whatever format is requested by the `set_config` for the output?

Reply:
To me these topics are very closely related:

- understanding feature names
- outputing things other than an `ndarray`
- accepting things other than `ndarray`

What I don't like here, is that we're making `ColumnTransformer` very specific, which doesn't make much sense to me. Why would the user need to use a `ColumnTransformer` in a `Pipeline` to be able to support this input? 

I think in my mind, the solution here would be in `_validate_data`, to read the feature names from this type, store them, and probably then convert it directly to a numpy array if needed, or a pandas DataFrame if we want to. I don't see why we should put this in `ColumnTransformer`.

Reply:
Ah, I see what you mean by having the logic to `_validate_data` and getting it to understand other types of DataFrames. I'll open another PR to update `_validate_data` to understand `feature_names_in_` from the dataframe protocol to get the ball rolling.

As an aside, with the recent addition of `cast_to_ndarray` to `_validate_data`, we can finally adapt `ColumnTransformer` to use `_validate_data` so it can reuse the logic.

Reply:
@thomasjpfan now that we are converging on #26411 and that more and more people express interest in other dataframes (e.g. #26435) it might be a good time to revive this PR (or the one you intended to open).

Reply:
I opened a PR for feature names here: https://github.com/scikit-learn/scikit-learn/pull/26464

As for this PR, I really want the ColumnTransformer to pass around the same DataFrame container as the input to the inner transformers. I suspect I still need https://github.com/data-apis/dataframe-api/issues/42 to work. Although, I can workaround it with defining a custom "from_dataframe" that is library specific and then update it when the DataFrame SPEc adds a "from_dataframe" functionality.

Reply:
+1 for a temporary library specific `from_dataframe` routing helper with a link to https://github.com/data-apis/dataframe-api/issues/42 in its docstring.